File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func helper(root *TreeNode, subRoot *TreeNode)bool{
	if root==nil && subRoot ==nil {
		return true
	}
	if root ==nil || subRoot ==nil{
		return false
	}
	
	

	if !(root.Val==subRoot.Val){
		return false
	}
	left:=helper(root.Left,subRoot.Left)
	right:=helper(root.Right,subRoot.Right)

	return left && right



}
func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	if root ==nil{return false}

	if root.Val==subRoot.Val{
		if helper(root,subRoot){
			return true
		}
	}
	left:=isSubtree(root.Left,subRoot)
	right:=isSubtree(root.Right,subRoot)

	return left || right 
	
    
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// class Solution {
// public:
//     int height(TreeNode* root){
//         if(!root) return 0;

//         return 1+max(height(root->right), height(root->left));
    
//     }
//     bool isBalanced(TreeNode* root) {
//         if(!root) return true;
//         bool bl = isBalanced(root->left);
//         bool br = isBalanced(root->right);
//         if(!bl || !br)return false;
//         int l=height(root->left);
//         int r=height(root->right);

//         if(abs(l-r)>1){
//             return false;
//         }
//         return true;
//     }
// };



class Solution {
public:
    int height(TreeNode* root){
        if(!root) return 0;
        int l=height(root->left);
        if (l == -1) return -1;

        int r=height(root->right);
        if (r == -1) return -1;

        if(abs(l-r)>1){
            return -1;
        }
        return 1+max(height(root->right), height(root->left));
    
    }
    bool isBalanced(TreeNode* root) {
        return height(root)!=-1; 
    }
};
